File: implementations/java/src/main/java/io/rcl/Ast.java
package io.rcl;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;

interface AstNode { String kind(); }

final class StringNode implements AstNode {
  final String value;
  StringNode(String v) { value = v; }
  public String kind() { return "string"; }
}

final class NumberNode implements AstNode {
  final double value;
  NumberNode(double v) { value = v; }
  public String kind() { return "number"; }
}

final class BooleanNode implements AstNode {
  final boolean value;
  BooleanNode(boolean v) { value = v; }
  public String kind() { return "boolean"; }
}

final class ArrayNode implements AstNode {
  final List<AstNode> elements = new ArrayList<>();
  public String kind() { return "array"; }
}

final class BlockNode implements AstNode {
  final String name;
  final String argument;
  final LinkedHashMap<String, AstNode> properties = new LinkedHashMap<>();
  final List<String> propertyOrder = new ArrayList<>();
  final List<BlockNode> blocks = new ArrayList<>();
  final List<BlockNode> namedBlocks = new ArrayList<>();

  BlockNode(String n, String a) { name = n; argument = a; }
  public String kind() { return "block"; }
}

File: implementations/swift/Sources/RCLParser/AST.swift
public enum Node: Equatable {
    case string(StringNode)
    case number(NumberNode)
    case boolean(BooleanNode)
    case array(ArrayNode)
    case block(BlockNode)

    public var kind: String {
        switch self {
        case .string: return "string"
        case .number: return "number"
        case .boolean: return "boolean"
        case .array: return "array"
        case .block: return "block"
        }
    }
}

public struct DocumentNode: Equatable {
    public let kind: String
    public let blocks: [BlockNode]
    public let rootValue: Node?

    public init(kind: String = "document", blocks: [BlockNode], rootValue: Node? = nil) {
        self.kind = kind
        self.blocks = blocks
        self.rootValue = rootValue
    }
}

public struct BlockNode: Equatable {
    public let kind: String
    public let name: String
    public let argument: String?
    public let properties: [String: Node]
    public let blocks: [String: BlockNode]
    public let namedBlocks: [BlockNode]

    public init(
        kind: String = "block",
        name: String,
        argument: String? = nil,
        properties: [String: Node] = [:],
        blocks: [String: BlockNode] = [:],
        namedBlocks: [BlockNode] = []
    ) {
        self.kind = kind
        self.name = name
        self.argument = argument
        self.properties = properties
        self.blocks = blocks
        self.namedBlocks = namedBlocks
    }
}

public struct StringNode: Equatable { public let kind: String = "string"; public let value: String }
public struct NumberNode: Equatable { public let kind: String = "number"; public let value: Double }
public struct BooleanNode: Equatable { public let kind: String = "boolean"; public let value: Bool }
public struct ArrayNode: Equatable { public let kind: String = "array"; public let elements: [Node] }

public struct ParseError: Error {
    public let message: String
    public let line: Int
    public let column: Int

    public init(_ message: String, line: Int, column: Int) {
        self.message = message
        self.line = line
        self.column = column
    }
}
